### PR TITLE
FedExP

### PR DESCRIPTION
**Pull request by Roeia99**
## Description
Implementation of FedExP in flower.

## Related issues/PRs
[FedExP](#2015)
 
## Checklist

- [ ]  Implement proposed change
- [ ]  Write tests
- [ ]  Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ]  Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [ ]  Make CI checks pass
- [ ]  Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel #contributions)

### CODE REVIEW

**Comment by jafermarq** [comment]
Hi @Roeia99 and @omarmoo5 , let me know when you'd like me to give an initial review to your work here.

**Comment by omarmoo5** [comment]
> Hi @Roeia99 and @omarmoo5 , let me know when you'd like me to give an initial review to your work here.

Certainly! Although the code is mostly done, we are investigating some issues with the model accuracy while replicating the paper results. We'll let you know once the code is ready for review.

File: baselines/fedexp/fedexp/client.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+class FlowerClient(fl.client.NumPyClient):
+    """Standard Flower client for CNN training."""
+
+    def __init__(
+        self,
+        cid: int,
+        net: torch.nn.Module,
+        train_loader: DataLoader,
+        device: str,
```
Comment (jafermarq) [inline baselines/fedexp/fedexp/client.py:26]:
```suggestion
```

Comment (jafermarq) [reply baselines/fedexp/fedexp/client.py:26]: Let's remove it and make the clients discover the device. see my next comment

File: baselines/fedexp/fedexp/client.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        num_epochs: int,
+        data_ratio,
+    ):  # pylint: disable=too-many-arguments
+        print(f"Initializing Client {cid}")
+        self.cid = cid
+        self.net = net
+        self.train_loader = train_loader
+        self.device = torch.device(device)
```
Comment (jafermarq) [inline baselines/fedexp/fedexp/client.py:34]:
```suggestion
        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```

Comment (jafermarq) [reply baselines/fedexp/fedexp/client.py:34]: If Flower's VirtualClientEngine is using GPUs (i.e. if `num_gpus>0`) then the clients will select device `cuda`, else `cpu`.

File: baselines/fedexp/fedexp/models.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    """
+    criterion = torch.nn.CrossEntropyLoss()
+    optimizer = torch.optim.SGD(
+        net.parameters(),
+        lr=hyperparams["eta_l"],
+        momentum=0,
+        weight_decay=hyperparams["weight_decay"],
+    )
+    net.train()
+    for _ in tqdm(range(epochs), desc="Local Training ..."):
```
Comment (jafermarq) [inline baselines/fedexp/fedexp/models.py:152]: I think it's better to remove `tqdm` from here since if multiple clients run in parallel it will lead to not-very-useful logs.... It's great to have `tqdm` for the server evaluation though.

File: baselines/fedexp/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    "Topic :: Scientific/Engineering :: Artificial Intelligence",
+    "Topic :: Scientific/Engineering :: Mathematics",
+    "Topic :: Software Development",
+    "Topic :: Software Development :: Libraries",
+    "Topic :: Software Development :: Libraries :: Python Modules",
+    "Typing :: Typed",
+]
+
+[tool.poetry.dependencies]
+python = ">=3.8.15, <3.12.0"  # don't change this
```
Comment (jafermarq) [inline baselines/fedexp/pyproject.toml:40]:
```suggestion
python = ">=3.10.0, <3.11.0"  # don't change this
```

Comment (jafermarq) [reply baselines/fedexp/pyproject.toml:40]: Since we pin py3.10 in the `torch` and `torchvision` packages.... (see my next comment)

File: baselines/fedexp/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+flwr = { extras = ["simulation"], version = "1.5.0" }
+hydra-core = "1.3.2" # don't change this
+
```
Comment (jafermarq) [inline baselines/fedexp/pyproject.toml:43]: I moved here the packages you indicated in the readme to be installed via `pip`. This will do the same thing. Please add them to the `pyproject.toml`:
```
torch = { url = "https://download.pytorch.org/whl/cu116/torch-1.13.1%2Bcu116-cp310-cp310-linux_x86_64.whl"}
torchvision = { url = "https://download.pytorch.org/whl/cu116/torchvision-0.14.1%2Bcu116-cp310-cp310-linux_x86_64.whl"}
```

File: baselines/fedexp/pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
```
Comment (jafermarq) [inline baselines/fedexp/pyproject.toml:44]: In addition, `tqdm` and `matplotlib` are needed. you can add them to the `pyproject.toml` by adding these lines (which is what would happen if you do instead `poetry add tqdm matplotlib` from the terminal:
```
tqdm = "4.66.1"
matplotlib = "3.8.0"
```

File: baselines/fedexp/pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+[tool.poetry.dev-dependencies]
+isort = "==5.11.5"
+black = "==23.1.0"
+docformatter = "==1.5.1"
+mypy = "==0.961"
```
Comment (jafermarq) [inline baselines/fedexp/pyproject.toml:49]:
```suggestion
mypy = "==1.4.1"
```

File: baselines/fedexp/fedexp/conf/base.yaml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+clients_per_round: 20
+num_epochs: 20
+batch_size: 50
+
+server_device: cuda
+client_device: cuda
+
+client_resources:
+  num_cpus: 1
+  num_gpus: 0.05
```
Comment (jafermarq) [inline baselines/fedexp/fedexp/conf/base.yaml:15]: It might be wise raising this a bit. I'm not sure if everyone out there can fit 20 (1/20=0.05) clients on the same machine. Testing on my machine i see each client requires ~1.5GB of VRAM. I'm thinking better raise this to `0.1` and leave a comment in the README's "Hardware Setup" paragraph giving some advice on hot to adjust it.

Comment (jafermarq) [reply baselines/fedexp/fedexp/conf/base.yaml:15]: But very glad to see a baseline packing many clients on the same GPU 💯 :
Screenshot 2023-09-25 at 22 14 36

**Review by jafermarq — COMMENTED** [review]
Hi @Roeia99,

Thanks for the updates for `FedExP`. I have done a first pass through your code, mostly focusing on (1) can I install the environment, (2) does the code run?, (3) and a general first look through the code.

The code runs! I left a few comments. Please reply to them if you have some questions, or mark them as resolved when you add the changes suggested. In addition to those:
* Could you please remove the `EXTENDED_README.md` ? that's mostly intended to guide you through the process of creating a baseline.
* Once you have your results and if you create some figures, please add them to a directory called `_static` (that should sit in the same directory as your `pyproject.toml`)

File: baselines/fedexp/fedexp/main.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        client_resources={
+            "num_cpus": cfg.client_resources.num_cpus,
+            "num_gpus": cfg.client_resources.num_gpus,
+        },
+        ray_init_args={
+            "ignore_reinit_error": True,
+            "include_dashboard": False,
+            "num_cpus": 12,
+            "num_gpus": 1,
+            "_memory": 100*1024*1024*1024,
```
Comment (jafermarq) [inline baselines/fedexp/fedexp/main.py:86]: Are all these necessary?

Comment (Roeia99) [reply baselines/fedexp/fedexp/main.py:86]: it was necessary for our server to detect the resources.

Comment (jafermarq) [reply baselines/fedexp/fedexp/main.py:86]: Ok. I'd say better not to leave this hardcoded  (i.e. `num_cpus`, `num_gpus`, `_memory`)because for the final version of your code because this will make the `simulation_start` fail if someone wants to run your `FedExp` but doesn't have access to that many resources. Similarly, if someone has more, their system can potentially be underutilised. 

Happy to leave this there until before merging with `main`.

**Comment by jafermarq** [comment]
@Roeia99 @omarmoo5, please let me know when I should do another review of your `FedExP` implementation.

**Comment by omarmoo5** [comment]
> @Roeia99 @omarmoo5, please let me know when I should do another review of your `FedExP` implementation.

Hi @jafermarq , we are ready for another review :)

File: baselines/fedexp/fedexp/models.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        loss = criterion(log_probs, labels)
+        loss.backward()
+        if hyperparams["use_gradient_clipping"]:
+            torch.nn.utils.clip_grad_norm_(
+                parameters=net.parameters(), max_norm=hyperparams["max_norm"]
+            )
+        optimizer.step()
+        counter += 1
+        if counter >= epochs:
+            break
```
Comment (jafermarq) [inline baselines/fedexp/fedexp/models.py:174]: This is not used anymore since you moved it to the client's `fit()`. Can we remove this then?

File: baselines/fedexp/fedexp/client.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+            for batch_idx, (images, labels) in enumerate(self.train_loader):
+                images, labels = images.to(self.device), labels.to(self.device)
+                if config["use_data_augmentation"]:
+                    transform_train = transforms.Compose(
+                        [
+                            transforms.RandomCrop(32, padding=4),
+                            transforms.RandomHorizontalFlip(),
+                        ]
+                    )
+                    images = transform_train(images)
```
Comment (jafermarq) [inline baselines/fedexp/fedexp/client.py:68]: having this in the training loop is a bit odd. See my suggestion to do this directly in the `load_datasets()` function, which also requires making a small change to the configs.

File: baselines/fedexp/fedexp/conf/base.yaml [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+  _target_: fedexp.models.ResNet18
+  num_classes: ${dataset_config.num_classes}
+
+hyperparams:
+  eta_l: 0.01     # local learning rate
+  epsilon: 0.001
+  decay : 0.998
+  weight_decay: 0.0001
+  max_norm: 10
+  use_data_augmentation: true
```
Comment (jafermarq) [inline baselines/fedexp/fedexp/conf/base.yaml:31]:
```suggestion
```

Comment (jafermarq) [reply baselines/fedexp/fedexp/conf/base.yaml:31]: moved to `dataset_config` above.

File: baselines/fedexp/fedexp/conf/base.yaml [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+server_device: cuda
+
+client_resources:
+  num_cpus: 1
+  num_gpus: 0.1
+
+dataset_config:
+  name: CIFAR10
+  num_classes: 10
+  alpha: 0.3
```
Comment (jafermarq) [inline baselines/fedexp/fedexp/conf/base.yaml:19]:
```suggestion
  alpha: 0.3
  use_data_augmentation: true
```

File: baselines/fedexp/README.md [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+
+
+## Expected Results
+
+```bash
+poetry run -m fedexp.main --multirun strategy.alogrithm=[fedexp,fedavg]
+```
+| FedExP | FedAvg |
+|:----:|:----:|
+|![FedExP CIFAR10](_static/FedExP_Cifar10_S0.png) | ![FedAvg CIFAR10](_static/FedAvg_Cifar10_S5.png) |
```
Comment (jafermarq) [inline baselines/fedexp/README.md:129]: Better to show these two in a single plot, so they are easier to compare. Please use grid lines and a title to the plot.

Comment (jafermarq) [reply baselines/fedexp/README.md:129]: Same comment for CIFAR-100

File: baselines/fedexp/fedexp/client.py (hunk 11 above)
Comment (jafermarq) [reply baselines/fedexp/fedexp/client.py:68]: Please remote the `if` statement.

File: baselines/fedexp/fedexp/dataset.py [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+        raise NotImplementedError
+
+    data_directory = f"./data/{config.name.lower()}/"
+    ds_path = f"{data_directory}train_{num_clients}_{config.alpha:.2f}.pkl"
+    trans_cifar = transforms.Compose(
+        [
+            transforms.ToTensor(),
+            transforms.Normalize(mean=[0.491, 0.482, 0.447], std=[0.247, 0.243, 0.262]),
+        ]
+    )
```
Comment (jafermarq) [inline baselines/fedexp/fedexp/dataset.py:124]:
```suggestion
    def get_transforms(is_train: bool):
        t_list = [
                transforms.ToTensor(),
                transforms.Normalize(mean=[0.491, 0.482, 0.447], std=[0.247, 0.243, 0.262]),
            ]
        if config.use_data_augmentation and is_train:
            t_list.append(transforms.RandomCrop(32, padding=4))
            t_list.append(transforms.RandomHorizontalFlip())

        return transforms.Compose(t_list)
```

Comment (jafermarq) [reply baselines/fedexp/fedexp/dataset.py:124]: a simple function that returns transforms for training/tests

File: baselines/fedexp/fedexp/dataset.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+            dataset_train,
+            num_clients,
+            config.alpha,
+            config.num_classes,
+            ds_path,
+            partition_equal,
+        )
+
+    dataset_test = Dataset(
+        data_directory, train=False, download=True, transform=trans_cifar
```
Comment (jafermarq) [inline baselines/fedexp/fedexp/dataset.py:142]:
```suggestion
        data_directory, train=False, download=True, transform=get_transforms(False)
```

File: baselines/fedexp/fedexp/dataset.py [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+            transforms.ToTensor(),
+            transforms.Normalize(mean=[0.491, 0.482, 0.447], std=[0.247, 0.243, 0.262]),
+        ]
+    )
+    try:
+        with open(ds_path, "rb") as file:
+            train_datasets = pickle.load(file)
+    except FileNotFoundError:
+        dataset_train = Dataset(
+            data_directory, train=True, download=True, transform=trans_cifar
```
Comment (jafermarq) [inline baselines/fedexp/fedexp/dataset.py:130]:
```suggestion
            data_directory, train=True, download=True, transform=get_transforms(True)
```

File: baselines/fedexp/_static/FedExP_Cifar100_S5.png
Comment (jafermarq) [inline baselines/fedexp/_static/FedExP_Cifar100_S5.png:1]: I believe this is an old plot when the training loop had a bug, let's remove it?

**Review by jafermarq — COMMENTED** [review]
Hi @Roeia99 @omarmoo5 @yehias21 ,

Thanks for taking the time to apply those changes. I have highlighted a few more. 

By looking at your `FedAvg` results, it seems to converge much much faster than in the `FedExp` paper.  For example, your `FedAvg` for CIFAR-10 at round ~100 is 70% while in the paper it seems to be more closer to ~55%. Similarly, for CIFAR-100 at round ~125 you get ~35% while the  paper shows ~25%. I'm comparing your plots in the `README.md` to those in Fig 3 in the paper. 

It could be there is a  bug in your FedAvg implementation. After fixing it, you can keep this result as `FedAvg-ours` or something like that if there is an explanation on why it works so much better than in the paper.

File: baselines/fedexp/_static/FedExP_Cifar100_S5.png
Comment (omarmoo5) [reply baselines/fedexp/_static/FedExP_Cifar100_S5.png:1]: I have removed it now, Thanks :)

**Comment by WilliamLindskog** [comment]
Hi @omarmoo5 

Just checking in here to see if there is anything we can support you with in order for you to implement this baseline?

**Comment by WilliamLindskog** [comment]
Hi @omarmoo5, 

Just checking in here. Do you have any updates to this baseline?

**Comment by WilliamLindskog** [comment]
Hi @Roeia99, 

Are you still interested in completing this baseline? If so, could you review the comments and changes requested? 

Best regards,
William
